rlp/split_test.go

```diff
@@ -0,0 +1,91 @@
+package rlp
+
+import (
+	"bytes"
+	"reflect"
+	"testing"
+)
+
+func TestCountValues(t *testing.T) {
+	tests := []struct {
+		input string
+		count int
+		err   error
+	}{
+		{"", 0, nil},
+		{"00", 1, nil},
+		{"80", 1, nil},
+		{"C0", 1, nil},
+		{"010203", 3, nil},
+		{"01C40607080902", 3, nil},
+		{
+			input: "A12000BF49F440A1CD0527E4D06E2765654C0F56452257516D793A9B8D604DCFDF2AB853F851808D10000000000000000000000000A056E81F171BCC55A6FF8345E692C0F86E5B48E01B996CADC001622FB5E363B421A0C5D2460186F7233C927E7DB2DCC703C0E500B653CA82273B7BFAD8045D85A470",
+			count: 2,
+			err:   nil,
+		},
+	}
+	for i, test := range tests {
+		count, err := CountValues(unhex(test.input))
+		if count != test.count {
+			t.Errorf("test %d: count mismatch, got %d want %d\ninput: %s", i, count, test.count, test.input)
+		}
+		if !reflect.DeepEqual(err, test.err) {
+			t.Errorf("test %d: err mismatch, got %q want %q\ninput: %s", i, err, test.err, test.input)
+		}
+	}
+}
+
+func TestSplit(t *testing.T) {
+	tests := []struct {
+		input     string
+		kind      Kind
+		val, rest string
+		err       error
+	}{
+		{input: "01FFFF", kind: Byte, val: "01", rest: "FFFF"},
+		{input: "80FFFF", kind: String, val: "", rest: "FFFF"},
+		{input: "C3010203", kind: List, val: "010203"},
+		{input: "C60607080902", err: ErrValueTooLarge, rest: "C60607080902"},
+		{input: "B800", err: ErrCanonSize, rest: "B800"},
+		{input: "B90000", err: ErrCanonSize, rest: "B90000"},
+		{input: "B90055", err: ErrCanonSize, rest: "B90055"},
+		{input: "BA0002FFFF", err: ErrCanonSize, rest: "BA0002FFFF"},
```

Comment:
This cannot be tested with small strings and lists because their size must be encoded in the smallest way possible. `0xB802FFFF` is invalid, the correct encoding would be `0x82FFFF`.